### DISCUSSION TITLE
Sanctum always return Unauthenticated while using API auth

### DISCUSSION THREAD

**Discussion by pierremedard** [laravel/framework#48930] [Q&A]
Hi everyone !

I've been trying to use Sanctum to authenticate my client for a Laravel API but it seems like i'm always unable to authenticate myself using postman. I've browsed the web for a while in search of an answer but it seems like i can't find a fix.
On my backend side, i've followed the docs, my login method looks like this:
```
public function login(LoginRequest $request): JsonResponse
{
    $data = $request->validated();
    $user = User::where('email', $data['email'])->firstOrFail();

    if (!Hash::check($data['password'], $user->password)) {
        return $this->returnError('Wrong Password.', 400);
    }

    $device_name = $request->device_name ?? 'postman';
    $user->tokens()->where('name', $device_name)->delete();
    $token = $user->createToken($device_name)->plainTextToken;

    return $this->returnSuccess([ 'token' => $token ], 200);
}
```
In the api.php routing file:
`Route::post('/logout', 'logout')->middleware('auth:sanctum');`
The kernel:
```
protected $middlewareGroups = [
    'web' => [
        \App\Http\Middleware\EncryptCookies::class,
        \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class,
        \Illuminate\Session\Middleware\StartSession::class,
        \Illuminate\View\Middleware\ShareErrorsFromSession::class,
        \App\Http\Middleware\VerifyCsrfToken::class,
        \Illuminate\Routing\Middleware\SubstituteBindings::class,
    ],

    'api' => [
        \Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
        \Illuminate\Routing\Middleware\ThrottleRequests::class.':api',
        \Illuminate\Routing\Middleware\SubstituteBindings::class,
    ],
];
```
For the config, in auth.php i have:
```
'guards' => [
    'web' => [
        'driver' => 'session',
        'provider' => 'users',
    ],
    'api' => [
        'driver' => 'token',
        'provider' => 'users',
    ],
],
```
In sanctum.php at the end of the file:
`'guard' =>  'api'`

And for the .htaccess:
```
# Handle Authorization Header
RewriteCond %{HTTP:Authorization} .
RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
```
According to my research, this should be enough for sanctum to be correctly configured on the back side. Now for Postman, I add the Accept: application/json, the Referer: http://localhost:8000, and in the Authorization tab, i select Bearer with the value 10|mO1ISY4Yy9ionnulfndxTIDslcH4naVP8R5Yvuj5daf08f51 (a token that my register API returns). And it doesn't seem to work. Also, when i remove the middleware for my logout route and try to return the header values, the Authorization token is not there, it seems like it has been removed at some point, even $request->bearerToken(); returns null. I tried using the Authorization tab of Postman but also setting it directly in the Header, can't get it to work.

Laravel: 10.x
Sanctum: ^3.3
PHP: 8.2
Apache2

Did i miss something during my configuration or in Postman ? Has anyone ever experienced that ?
Thanks !

**Comment by Soburr** [comment]
I think it should be 
'api' => [
        'driver' => 'sanctum',
        'provider' => 'users',
    ],

**Reply by Sharifur** [reply]
tried but not working

**Comment by hoainam1402** [comment]
```
protected function authenticate($request, array $guards)
    {
        $token = PersonalAccessToken::findToken($request->bearerToken());
        // Check if the token is expired
        if ($token && $token->expires_at && ($token->expires_at < now())) {
            $token->delete();
            throw new ApExpiredTokenException();
        }

        parent::authenticate($request, $guards);
    }
```
my solution you need overwrite class Authenticate

**Comment by darkvovich** [comment]
Have the same problem in Laravel 10.x

Sanctum token is Valid, but return 401

**Reply by Sharifur** [reply]
@darkvovich ensure that you have set sanctum domain
SESSION_DOMAIN=localhost
SANCTUM_STATEFUL_DOMAINS=localhost,*.localhost

also you are are using php artisan serve for laravel backend

**Comment by dmcchanaka** [comment]
Same issue with laravel 12. Only receiving  401 in production server.

**Reply by elishamutang** [reply]
same here...anyone found solution?

**Reply by Yassine-Light** [reply]
i think i found it, you just need to change you auth.php config file
change the guard from web to sanctum, and define it below with the guard
    'defaults' => [
        'guard' => env('AUTH_GUARD', 'sanctum'),
        'passwords' => env('AUTH_PASSWORD_BROKER', 'users'),
    ],
    
    
        'guards' => [
        'web' => [
            'driver' => 'session',
            'provider' => 'users',
        ],
        'sanctum' => [
            'driver' => 'sanctum',
            'provider' => 'users',
        ],
    ],
    
    and clear your config 
    php artisan config:clear
php artisan cache:clear
php artisan route:clear
php artisan optimize:clear

and change the .htaccess file  in the handle authorization to this 
    # Handle Authorization Header
    RewriteCond %{HTTP:Authorization} ^(.+)$
    RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]

**Reply by elishamutang** [reply]
> i think i found it, you just need to change you auth.php config file
> 
> change the guard from web to sanctum, and define it below with the guard
> 
>     'defaults' => [
> 
>         'guard' => env('AUTH_GUARD', 'sanctum'),
> 
>         'passwords' => env('AUTH_PASSWORD_BROKER', 'users'),
> 
>     ],
> 
>     
> 
>     
> 
>         'guards' => [
> 
>         'web' => [
> 
>             'driver' => 'session',
> 
>             'provider' => 'users',
> 
>         ],
> 
>         'sanctum' => [
> 
>             'driver' => 'sanctum',
> 
>             'provider' => 'users',
> 
>         ],
> 
>     ],
> 
>     
> 
>     and clear your config 
> 
>     php artisan config:clear
> 
> php artisan cache:clear
> 
> php artisan route:clear
> 
> php artisan optimize:clear
> 
> 
> 
> and change the .htaccess file  in the handle authorization to this 
> 
>     # Handle Authorization Header
> 
>     RewriteCond %{HTTP:Authorization} ^(.+)$
> 
>     RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
> 
> 

Does this still work while having the normal web auth as well? Last time I tried this but my web auth wasn't working. For my use case I needed both web and sanctum to work. In the end I implemented a custom header that contained the bearer token and checked for this in a custom middleware on each request.

**Comment by alphaisx** [comment]
If using wamp, the issue is with the header not being delivered to localhost

In .htaccess, add at the top:
```
RewriteEngine On

RewriteCond %{HTTP:Authorization} .
RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
```
